src/index.js

```diff
@@ -0,0 +1,194 @@
+          'that is the signal that it should be rendering its part of the UI. It may in turn decide not to',
+          'render all of *its* children, but it should never return `null` for itself. It is not responsible',
+          'for that decision.'
+        ].join(' '));
+      }
+
+      const fn = this.__outsideClickHandler = generateOutsideCheck(
+        componentNode,
+        instance,
+        clickOutsideHandler,
+        this.props.outsideClickIgnoreClass,
+        this.props.excludeScrollbar,
+        this.props.preventDefault,
+        this.props.stopPropagation
+      );
+
+      const pos = registeredComponents.length;
+      registeredComponents.push(this);
+      handlers[pos] = fn;
+
+      // If there is a truthy disableOnClickOutside property for this
+      // component, don't immediately start listening for outside events.
+      if (!this.props.disableOnClickOutside) {
+        this.enableOnClickOutside();
+      }
+    }
+
+    /**
+    * Track for disableOnClickOutside props changes and enable/disable click outside
+    */
+    componentWillReceiveProps(nextProps) {
+      if (this.props.disableOnClickOutside && !nextProps.disableOnClickOutside) {
+        this.enableOnClickOutside();
+      } else if (!this.props.disableOnClickOutside && nextProps.disableOnClickOutside) {
+        this.disableOnClickOutside();
+      }
+    }
+
+    /**
+     * Remove the document's event listeners
+     */
+    componentWillUnmount() {
+      this.disableOnClickOutside();
+      this.__outsideClickHandler = false;
+      const pos = registeredComponents.indexOf(this);
+      if (pos > -1) {
+        // clean up so we don't leak memory
+        if (handlers[pos]) { handlers.splice(pos, 1); }
+        registeredComponents.splice(pos, 1);
+      }
+    }
+
+    /**
+     * Can be called to explicitly enable event listening
+     * for clicks and touches outside of this element.
+     */
+    enableOnClickOutside = () => {
+      const fn = this.__outsideClickHandler;
+      if (typeof document !== 'undefined') {
+        let events = this.props.eventTypes;
+        if (!events.forEach) {
+          events = [events];
+        }
+        events.forEach(eventName => document.addEventListener(eventName, fn));
+      }
+    }
+
+    /**
+     * Can be called to explicitly disable event listening
+     * for clicks and touches outside of this element.
+     */
+    disableOnClickOutside = () => {
+      const fn = this.__outsideClickHandler;
+      if (typeof document !== 'undefined') {
+        let events = this.props.eventTypes;
+        if (!events.forEach) {
+          events = [events];
+        }
+        events.forEach(eventName => document.removeEventListener(eventName, fn));
+      }
+    }
+
+    getRef = ref => this.instanceRef = ref
+
+    /**
+     * Pass-through render
+     */
+    render() {
+      var props = Object.keys(this.props)
```

Comment:
hmm, we can probably put this rewriting in the willMount and willUpdate functions, so that we cache the result of this filter/reduce and property extension on incoming props as, say, `this.__cached_props` or something and then pass those on during render instead, so we don't rebuild an invariable during render